core/templates/dev/head/components/forms/FormBuilderSpec.js

```diff
@@ -146,6 +169,212 @@ describe('Normalizer tests', function() {
+
+  it('should invalidate if value is not a float', function() {
+    testInput.$setViewValue('-abc');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+
+    testInput.$setViewValue('3..5');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+
+    testInput.$setViewValue('-2.abc');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+
+    testInput.$setViewValue('0.3.5');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+
+    testInput.$setViewValue(undefined);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+  });
+});
+
+describe('Testing apply-validation directive', function() {
+  var $compile, element, scope, testInput;
+
+  beforeEach(module('oppia'));
+
+  beforeEach(inject(function($rootScope) {
+    scope = $rootScope.$new();
+    element = '<form name="testForm">' +
+      '<input name="inputValue" type="number" ng-model="localValue" ' +
+      'apply-validation>' +
+      '</form>';
+  }));
+
+  it('should apply isAtLeast validation', inject(function($compile) {
+    scope.validators = function() {
+      return [{
+        id: 'isAtLeast',
+        minValue: -2.5
+      }];
+    };
+    $compile(element)(scope);
+    testInput = scope.testForm.inputValue;
+
+    testInput.$setViewValue(-1);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue('1');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue(-3);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+
+    testInput.$setViewValue('-3');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+  }));
+
+  it('should apply isAtMost validation', inject(function($compile) {
+    scope.validators = function() {
+      return [{
+        id: 'isAtMost',
+        maxValue: 5
+      }];
+    };
+    $compile(element)(scope);
+    testInput = scope.testForm.inputValue;
+
+    testInput.$setViewValue(-1);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue('1');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue(6);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+
+    testInput.$setViewValue('10');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+  }));
+
+  it('should apply isNonempty validation', inject(function($compile) {
+    scope.validators = function() {
+      return [{
+        id: 'isNonempty'
+      }];
+    };
+    $compile(element)(scope);
+    testInput = scope.testForm.inputValue;
+
+    testInput.$setViewValue(-1);
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue('1');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(true);
+
+    testInput.$setViewValue('');
+    scope.$digest();
+    expect(testInput.$valid).toEqual(false);
+
+    testInput.$setViewValue(undefined);
```

Comment:
This one is probably not going to happen in practice, I think? (If that's the case, no harm adding it here, but also probably not necessary IMO.)